### ISSUE TITLE
Consuming the winget WinRT COM API via windows-rs?

### ISSUE THREAD

**Issue by fgimian** [microsoft/windows-rs#3352]
### Summary

Hey there Kenny and friends, I may be missing something here so not entirely sure if this is a bug or simply something I'm doing wrong.

Here's an extremely minimal C# project which gets one property from one interface.

**ConsoleApp1.csproj**

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net9.0-windows10.0.26100.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Windows.CsWin32" Version="0.3.106">
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
    <PackageReference Include="Microsoft.Windows.CsWinRT" Version="2.2.0" />
    <PackageReference Include="Microsoft.WindowsPackageManager.InProcCom" Version="1.9.25180">
      <NoWarn>NU1701</NoWarn>
      <GeneratePathProperty>true</GeneratePathProperty>
      <IncludeAssets>none</IncludeAssets>
    </PackageReference>
  </ItemGroup>

  <PropertyGroup>
    <CsWinRTIncludes>Microsoft.Management.Deployment</CsWinRTIncludes>
  </PropertyGroup>

  <ItemGroup>
    <CsWinRTInputs Include="$(TargetDir)\Microsoft.Management.Deployment.winmd" />
  </ItemGroup>

  <Target Name="CopyWinmdToTargetDir" BeforeTargets="BeforeBuild">
    <Copy SourceFiles="$(PkgMicrosoft_WindowsPackageManager_InProcCom)\lib\Microsoft.Management.Deployment.winmd" DestinationFolder="$(TargetDir)" />
  </Target>

</Project>
```

**Program.cs**

```cs
using System.Runtime.InteropServices;
using Microsoft.Management.Deployment;
using Windows.Win32;
using Windows.Win32.System.Com;

var hr = PInvoke.CoCreateInstance(
    rclsid: new Guid("526534B8-7E46-47C8-8416-B1685C327D37"),
    pUnkOuter: null,
    dwClsContext: CLSCTX.CLSCTX_LOCAL_SERVER,
    riid: typeof(ICreateCompositePackageCatalogOptions).GUID,
    ppv: out object result
);
Marshal.ThrowExceptionForHR(hr);

var pUnknown = Marshal.GetIUnknownForObject(result);
var options = WinRT.MarshalGeneric<CreateCompositePackageCatalogOptions>.FromAbi(pUnknown);
if (pUnknown != nint.Zero)
{
    Marshal.Release(pUnknown);
}

Console.WriteLine(options.InstalledScope);
```

The output is the default install scope:

```
Any
```

Now, in Rust, I've grabbed a copy of the winmd file and copied it to a **winmd** sub-directory. The Cargo.toml file and all source code is found below in the specific sections of the issue template.

The output is the following error:

```
Error: Error { code: HRESULT(0x80073D54), message: "Failed to find proxy registration for IID: {21ABAA76-089D-51C5-A745-C85EEFE70116}." }
```

### Crate manifest

```TOML
[package]
name = "winget-example"
version = "0.1.0"
edition = "2021"

[dependencies]
windows-core = "0.58.0"

[dependencies.windows]
version = "0.58.0"
features = [
    "Foundation_Collections",
    "System",
    "Win32_System_Com",
]

[build-dependencies]
windows-bindgen = "0.58.0"
```

### Crate code

**build.rs**

```rust
fn main() {
    if !cfg!(target_env = "msvc") {
        return;
    }

    println!("cargo:rustc-link-lib=onecoreuap");

    windows_bindgen::bindgen([
        "--in",
        "winmd/Microsoft.Management.Deployment.winmd",
        "--out",
        "src/bindings.rs",
        "--filter",
        "Microsoft.Management.Deployment",
        "--config",
        "no-bindgen-comment",
    ])
    .unwrap();
}
```

**main.rs**

```rust
#![cfg(target_env = "msvc")]
mod bindings;

use windows::Win32::System::Com::{CoCreateInstance, CoInitialize, CLSCTX_LOCAL_SERVER};
use windows_core::{IUnknown, Interface, GUID};

use crate::bindings::Management::Deployment::CreateCompositePackageCatalogOptions;

fn main() -> windows_core::Result<()> {
    let clsid = GUID::from_u128(0x526534B8_7E46_47C8_8416_B1685C327D37);

    unsafe {
        CoInitialize(None).ok()?;
        let options: IUnknown = CoCreateInstance(&clsid, None, CLSCTX_LOCAL_SERVER)?;
        let options: CreateCompositePackageCatalogOptions = options.cast()?;
        println!("{:?}", options.InstalledScope()?);
    }

    Ok(())
}
```

**Comment by riverar** [comment]
Neither the C# project or the Rust crate compile, feel free to attach a zip if that'll help.

**Comment by fgimian** [comment]
Sure, so you'll first need to install the [Windows SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/).

And here are all files from the C# project, I did forget to mention NativeMethods.txt, sorry 😄 

[ConsoleApp1.zip](https://github.com/user-attachments/files/17930023/ConsoleApp1.zip)

Please let me know if that runs for you.

Cheers
Fotis

**Comment by riverar** [comment]
Microsoft.DesktopAppInstaller is utilizing Packaged COM; let me chat with Kenny about our level of support for that and get back to you.

**Comment by fgimian** [comment]
> Microsoft.DesktopAppInstaller is utilizing Packaged COM; let me chat with Kenny about our level of support for that and get back to you.

Awesome, thank you so much Rafael, really appreciate it! 😄

**Comment by riverar** [comment]
tl;dr: Ensure `Microsoft.Management.Deployment.winmd` is next to your executable.

Because the interface isn’t registered in any catalog, the system looks for metadata (`Microsoft.Management.Deployment.winmd`) to determine how to handle marshaling for the interface. This metadata specifies the marshaling behavior (in this case, _agile_) and other activation details, allowing the system to set up cross-process communication correctly.

**Comment by riverar** [comment]
Feel free to continue the discussion or let us know how it went!

**Comment by fgimian** [comment]
> tl;dr: Ensure `Microsoft.Management.Deployment.winmd` is next to your executable.
> 
> Because the interface isn’t registered in any catalog, the system looks for metadata (`Microsoft.Management.Deployment.winmd`) to determine how to handle marshaling for the interface. This metadata specifies the marshaling behavior (in this case, _agile_) and other activation details, allowing the system to set up cross-process communication correctly.

I honestly don't know why I didn't think of that since I'm doing this in the C# version 😆 

> Feel free to continue the discussion or let us know how it went!

It works perfectly! Here's the final **build.rs** which copies the winmd file to the build directory:

```rust
use std::{env, fs, path::Path};

fn main() {
    if !cfg!(target_env = "msvc") {
        return;
    }

    println!("cargo:rustc-link-lib=onecoreuap");

    windows_bindgen::bindgen([
        "--in",
        "winmd/Microsoft.Management.Deployment.winmd",
        "--out",
        "src/bindings.rs",
        "--filter",
        "Microsoft.Management.Deployment",
        "--config",
        "no-bindgen-comment",
    ])
    .unwrap();

    let out_dir = env::var("OUT_DIR").unwrap();
    let out_dir = Path::new(&out_dir);

    let binary_directory_path = out_dir
        .parent()
        .and_then(|path| path.parent())
        .and_then(|path| path.parent())
        .unwrap_or(out_dir);

    fs::copy(
        "winmd/Microsoft.Management.Deployment.winmd",
        binary_directory_path.join("Microsoft.Management.Deployment.winmd"),
    )
    .unwrap();
}
```

And the working **main.rs**:

```rust
#![cfg(target_env = "msvc")]
mod bindings;

use windows::Win32::System::Com::{CoCreateInstance, CoInitialize, CLSCTX_LOCAL_SERVER};
use windows_core::{IUnknown, Interface, GUID};

use crate::bindings::Management::Deployment::CreateCompositePackageCatalogOptions;

fn main() -> windows_core::Result<()> {
    let clsid = GUID::from_u128(0x526534B8_7E46_47C8_8416_B1685C327D37);

    unsafe {
        CoInitialize(None).ok()?;
        let options: IUnknown = CoCreateInstance(&clsid, None, CLSCTX_LOCAL_SERVER)?;
        let options = options.cast::<CreateCompositePackageCatalogOptions>()?;
        println!("{:?}", options.InstalledScope()?);
    };

    Ok(())
}
```

Thank you so so much for your help with this!
Fotis

**Comment by fgimian** [comment]
Here's a more complete example which successfully connects to the local package manager and searches for a package.

```rust
#![cfg(target_env = "msvc")]
mod bindings;

use windows::Win32::System::Com::{CoCreateInstance, CoInitialize, CLSCTX_LOCAL_SERVER};
use windows_core::{h, IUnknown, Interface, GUID};

use crate::bindings::Management::Deployment::{
    CompositeSearchBehavior, ConnectResultStatus, CreateCompositePackageCatalogOptions,
    FindPackagesOptions, FindPackagesResultStatus, PackageFieldMatchOption, PackageInstallScope,
    PackageManager, PackageMatchField, PackageMatchFilter,
};

// Define all required CSLIDs for package manager interfaces.
static PACKAGE_MANAGER_CLSID: GUID = GUID::from_u128(0xC53A4F16_787E_42A4_B304_29EFFB4BF597);
static FIND_PACKAGES_OPTIONS_CLSID: GUID = GUID::from_u128(0x572DED96_9C60_4526_8F92_EE7D91D38C1A);
static CREATE_COMPOSITE_PACKAGE_CATALOG_OPTIONS_CLSID: GUID =
    GUID::from_u128(0x526534B8_7E46_47C8_8416_B1685C327D37);
static INSTALL_OPTIONS_CLSID: GUID = GUID::from_u128(0x1095F097_EB96_453B_B4E6_1613637F3B14);
static UNINSTALL_OPTIONS_CLSID: GUID = GUID::from_u128(0xE1D9A11E_9F85_4D87_9C17_2B93143ADB8D);
static PACKAGE_MATCH_FILTER_CLSID: GUID = GUID::from_u128(0xD02C9DAF_99DC_429C_B503_4E504E4AB000);

fn main() -> windows_core::Result<()> {
    unsafe {
        CoInitialize(None).ok()?;
    }

    // Create an instance of the package manager.
    let package_manager = create_standard_instance::<PackageManager>(&PACKAGE_MANAGER_CLSID)?;

    // Obtain catalogs for the Microsoft Store and winget.
    let msstore_catalog = package_manager.GetPackageCatalogByName(h!("msstore"))?;
    let winget_catalog = package_manager.GetPackageCatalogByName(h!("winget"))?;

    // Configure catalog options to include both catalogs and search locally installed packages
    // defined at any scope.
    let composite_catalog_options = create_standard_instance::<CreateCompositePackageCatalogOptions>(
        &CREATE_COMPOSITE_PACKAGE_CATALOG_OPTIONS_CLSID,
    )?;
    composite_catalog_options
        .Catalogs()?
        .Append(&msstore_catalog)?;
    composite_catalog_options
        .Catalogs()?
        .Append(&winget_catalog)?;
    composite_catalog_options.SetCompositeSearchBehavior(CompositeSearchBehavior::LocalCatalogs)?;
    composite_catalog_options.SetInstalledScope(PackageInstallScope::Any)?;

    // Connect to the package manager using the options defined above.
    let connect_result = package_manager
        .CreateCompositePackageCatalog(&composite_catalog_options)?
        .Connect()?;
    if connect_result.Status()? != ConnectResultStatus::Ok {
        panic!("failed to connect to the package catalog");
    }

    // Create find options for performing a search.
    let find_packages_options =
        create_standard_instance::<FindPackagesOptions>(&FIND_PACKAGES_OPTIONS_CLSID)?;

    // Add a filter to search for a package with the ID Git.Git.
    let match_filter = create_standard_instance::<PackageMatchFilter>(&PACKAGE_MATCH_FILTER_CLSID)?;
    match_filter.SetField(PackageMatchField::Id)?;
    match_filter.SetOption(PackageFieldMatchOption::Equals)?;
    match_filter.SetValue(h!("Git.Git"))?;
    find_packages_options.Selectors()?.Append(&match_filter)?;

    // Perform the search.
    let find_result = connect_result
        .PackageCatalog()?
        .FindPackages(&find_packages_options)?;
    if find_result.Status()? != FindPackagesResultStatus::Ok {
        panic!("failed to find packages");
    }

    // Display all matches.
    for result in find_result.Matches()? {
        let catalog_package = result.CatalogPackage()?;

        // Verify that the package originates from a known package catalog and obtain its name.
        let Ok(default_install_version) = catalog_package.DefaultInstallVersion() else {
            continue;
        };

        let catalog_name = default_install_version.PackageCatalog()?.Info()?.Name()?;
        if catalog_name.is_empty() {
            continue;
        }

        println!(
            "{} - {} - {catalog_name}",
            catalog_package.Id()?,
            catalog_package.Name()?
        );
    }

    Ok(())
}

fn create_standard_instance<T: windows_core::Interface>(clsid: &GUID) -> windows_core::Result<T> {
    unsafe {
        let object: IUnknown = CoCreateInstance(clsid, None, CLSCTX_LOCAL_SERVER)?;
        object.cast::<T>()
    }
}

// TODO: Use the WinGetServerManualActivation_CreateInstance function from winrtact.dll to create
// an instance when the user is running as Administrator.
// fn create_elevated_instance<T: windows_core::Interface>(clsid: &GUID) -> windows_core::Result<T> {}
```

Output is as follows:

```
Git.Git - Git - winget
```

I can't thank you enough for helping me with this, I honestly never thought I would get it working.

Cheers
Fotis
